File: backend/app/services/astrology_service.py
# Este archivo contiene la lógica para calcular la carta natal básica utilizando la biblioteca pyswisseph.
# Se incluyen funciones para convertir la fecha y hora de nacimiento a UTC, 
# obtener la zona horaria a partir de las coordenadas, 
# calcular la longitud eclíptica de los planetas
from datetime import datetime
from zoneinfo import ZoneInfo

import swisseph as swe
from timezonefinder import TimezoneFinder


ZODIAC_SIGNS = [
    "Aries",
    "Tauro",
    "Géminis",
    "Cáncer",
    "Leo",
    "Virgo",
    "Libra",
    "Escorpio",
    "Sagitario",
    "Capricornio",
    "Acuario",
    "Piscis",
]

SIGN_ELEMENTS = {
    "Aries": "Fuego",
    "Leo": "Fuego",
    "Sagitario": "Fuego",
    "Tauro": "Tierra",
    "Virgo": "Tierra",
    "Capricornio": "Tierra",
    "Géminis": "Aire",
    "Libra": "Aire",
    "Acuario": "Aire",
    "Cáncer": "Agua",
    "Escorpio": "Agua",
    "Piscis": "Agua",
}

# Función para obtener el signo zodiacal a partir de la longitud eclíptica del planeta
def get_zodiac_sign_from_longitude(longitude: float) -> str:
    normalized_longitude = longitude % 360
    sign_index = int(normalized_longitude // 30)
    return ZODIAC_SIGNS[sign_index]

# Función para obtener la zona horaria a partir de las coordenadas
def get_timezone_desde_coordenadas(lat: float, lon: float) -> str:
    tf = TimezoneFinder()
    timezone_nombre = tf.timezone_at(lat=lat, lng=lon)

    if timezone_nombre is None:
        raise ValueError("No se pudo determinar la zona horaria del lugar de nacimiento.")

    return timezone_nombre

# Función para convertir la fecha y hora de nacimiento a UTC
def get_utc_fechahora(fecha_nacimiento: str, birth_time: str, lat: float, lon: float) -> datetime:
    timezone_nombre = get_timezone_desde_coordenadas(lat, lon)

    local_datetime = datetime.fromisoformat(f"{fecha_nacimiento}T{birth_time}")
    local_datetime = local_datetime.replace(tzinfo=ZoneInfo(timezone_nombre))

    return local_datetime.astimezone(ZoneInfo("UTC"))

# Función para calcular el signo zodiacal de un planeta dado el día juliano y el planeta
def calculate_planet_sign(julian_day: float, planet: int) -> str:
    result = swe.calc_ut(julian_day, planet)

    # pyswisseph devuelve una tupla. El primer elemento contiene los datos del planeta.
    longitude = result[0][0]

    return get_zodiac_sign_from_longitude(longitude)

# Función para calcular el signo ascendente usando casas Placidus.
def calcular_signo_ascendente(julian_day: float, lat: float, lon: float) -> str:
    _, ascmc = swe.houses(julian_day, lat, lon, b"P")

    # En pyswisseph, ascmc[0] corresponde a la longitud zodiacal del ascendente.
    asc_longitude = ascmc[0]

    return get_zodiac_sign_from_longitude(asc_longitude)

# Función principal para calcular el mapa natal básico
def calculo_carta_basica(
    fecha_nacimiento: str,
    birth_time: str,
    lat: float,
    lon: float
) -> dict:
    utc_datetime = get_utc_fechahora(fecha_nacimiento, birth_time, lat, lon)

    utc_hour = (
        utc_datetime.hour
        + utc_datetime.minute / 60
        + utc_datetime.second / 3600
    )

    julian_day = swe.julday(
        utc_datetime.year,
        utc_datetime.month,
        utc_datetime.day,
        utc_hour
    )

    sol_sign = calculate_planet_sign(julian_day, swe.SUN)
    luna_sign = calculate_planet_sign(julian_day, swe.MOON)
    asc_sign = calcular_signo_ascendente(julian_day, lat, lon)

    return {
        "solSign": sol_sign,
        "lunaSign": luna_sign,
        "ascSign": asc_sign,
        "elements": {
            "sun": SIGN_ELEMENTS[sol_sign],
            "moon": SIGN_ELEMENTS[luna_sign],
            "asc": SIGN_ELEMENTS[asc_sign],
        }
    }

File: frontend/src/services/astrologyService.ts

import { SwissEphemeris, Planet } from "@swisseph/browser"
import type { ZodiacSign } from "../types/astrology"


let sweInstance: SwissEphemeris | null = null
async function getSwissEphemeris() {
  if (!sweInstance) {
    sweInstance = new SwissEphemeris()
    await sweInstance.init()
  }

  return sweInstance
}

export function getZodiacSignFromLongitude(longitude: number): ZodiacSign {

    if (!Number.isFinite(longitude)) {
    throw new Error("Debe ser un numero válido.")
    }    
    const zodiacSigns: ZodiacSign[] = [
        "Aries",
        "Tauro",
        "Géminis",
        "Cáncer",
        "Leo",
        "Virgo",
        "Libra",
        "Escorpio",
        "Sagitario",
        "Capricornio",
        "Acuario",
        "Piscis",
    ]

    const normalizedLongitude = ((longitude % 360) + 360) % 360
    const signIndex = Math.floor(normalizedLongitude / 30)

    return zodiacSigns[signIndex]
}

export function getSunSign(birthDate: string): ZodiacSign {
    const date = new Date(`${birthDate}T12:00:00`)
    const month = date.getMonth() + 1
    const day = date.getDate()

    if ((month === 3 && day >= 21) || (month === 4 && day <= 19)) return "Aries"
    if ((month === 4 && day >= 20) || (month === 5 && day <= 20)) return "Tauro"
    if ((month === 5 && day >= 21) || (month === 6 && day <= 20)) return "Géminis"
    if ((month === 6 && day >= 21) || (month === 7 && day <= 22)) return "Cáncer"
    if ((month === 7 && day >= 23) || (month === 8 && day <= 22)) return "Leo"
    if ((month === 8 && day >= 23) || (month === 9 && day <= 22)) return "Virgo"
    if ((month === 9 && day >= 23) || (month === 10 && day <= 22)) return "Libra"
    if ((month === 10 && day >= 23) || (month === 11 && day <= 21)) return "Escorpio"
    if ((month === 11 && day >= 22) || (month === 12 && day <= 21)) return "Sagitario"
    if ((month === 12 && day >= 22) || (month === 1 && day <= 19)) return "Capricornio"
    if ((month === 1 && day >= 20) || (month === 2 && day <= 18)) return "Acuario"

    // si no es ninguno anterior, devuelve picis
    return "Piscis"
}

export async function getMoonSign(
  birthDate: string,
  birthTime: string
): Promise<ZodiacSign> {
  const swe = await getSwissEphemeris()

  const date = new Date(`${birthDate}T${birthTime}:00`)
  const year = date.getFullYear()
  const month = date.getMonth() + 1
  const day = date.getDate()

  const hour =
    date.getHours() +
    date.getMinutes() / 60 +
    date.getSeconds() / 3600

  const jd = swe.julianDay(year, month, day, hour)
  const moon = swe.calculatePosition(jd, Planet.Moon)

  return getZodiacSignFromLongitude(moon.longitude)
}
